wrapper/src/main/java/software/amazon/jdbc/plugin/customendpoint/CustomEndpointMonitorImpl.java

```diff
@@ -0,0 +1,229 @@
+    } finally {
+      this.pluginServicesLock.writeLock().unlock();
+    }
+  }
+
+  @Override
+  public boolean shouldDispose() {
+    this.pluginServicesLock.readLock().lock();
+    try {
+      return this.pluginServices.isEmpty();
+    } finally {
+      this.pluginServicesLock.readLock().unlock();
+    }
+  }
+
+  @Override
+  public void close() {
+    this.stop.set(true);
+
+    try {
+      // TODO: the termination timeout is taken from failover2, should it be shorter since it blocks the current thread?
+      if (!this.monitorExecutor.awaitTermination(30, TimeUnit.SECONDS)) {
+        this.monitorExecutor.shutdownNow();
+        this.rdsClient.close();
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      this.monitorExecutor.shutdownNow();
+      this.rdsClient.close();
+    }
+  }
+
+  @Override
+  public void run() {
+    try {
+      while (!this.stop.get() && !Thread.currentThread().isInterrupted()) {
+        long start = System.nanoTime();
+
+        final Filter customEndpointFilter =
+            Filter.builder().name("db-cluster-endpoint-type").values("custom").build();
+        final DescribeDbClusterEndpointsResponse endpointsResponse =
+            this.rdsClient.describeDBClusterEndpoints(
+                (builder) ->
+                    builder.dbClusterEndpointIdentifier(this.endpointIdentifier).filters(customEndpointFilter));
+        List<DBClusterEndpoint> endpoints = endpointsResponse.dbClusterEndpoints();
+        if (endpoints.size() != 1) {
+          List<String> endpointURLs = endpoints.stream().map(DBClusterEndpoint::endpoint).collect(Collectors.toList());
+          LOGGER.warning(
+              Messages.get("CustomEndpointMonitorImpl.unexpectedNumberOfEndpoints",
+                  new Object[] {
+                      this.endpointIdentifier,
+                      this.region.id(),
+                      endpoints.size(),
+                      endpointURLs
+                  }
+              ));
+
+          // TODO: should we throw an exception and then return? Or wait and then try again?
+          long elapsedTime = System.nanoTime() - start;
+          long sleepDuration = Math.min(0, this.refreshRateNano - elapsedTime);
+          TimeUnit.NANOSECONDS.sleep(sleepDuration);
+          continue;
+        }
+
+        CustomEndpointInfo endpointInfo = CustomEndpointInfo.fromDBClusterEndpoint(endpoints.get(0));
+        if (customEndpointInfoCache.get(this.endpointIdentifier).equals(endpointInfo)) {
+          long elapsedTime = System.nanoTime() - start;
+          long sleepDuration = Math.min(0, this.refreshRateNano - elapsedTime);
+          TimeUnit.NANOSECONDS.sleep(sleepDuration);
+          continue;
+        }
+
+        // The custom endpoint info has changed, so we need to update the info in the registered plugin services.
+        customEndpointInfoCache.put(this.endpointIdentifier, endpointInfo, this.cacheEntryExpirationNano);
```

Comment:
Another thread might add such item in the cache. Would it be damaging? Can we use computeIfAbsent()?

Reply:
I don't think any other thread will be placing info at this key, the logic in CustomEndpointPlugin should only be creating one thread per custom endpoint which is shared among all connections, which means there is only one thread updating info for a given custom endpoint. computeIfAbsent would allow us to place an initial custom endpoint info object in the cache but would not allow us to update the info later.